File: scrape_readmes.py
#!/usr/bin/env python3
"""
Scrapes README.md files for top 1,000 repositories using GitHub API (gh auth token or HTTPS),
stores markdown files in `readmes/`, and indexes them in SQLite FTS5 database `data/repos.db`.
"""

import concurrent.futures
import json
import os
import sqlite3
import subprocess
import sys
import time
import urllib.request

SCRIPT_DIR = os.path.dirname(os.path.abspath(__file__))
READMES_DIR = os.path.join(SCRIPT_DIR, "readmes")
DATA_DIR = os.path.join(SCRIPT_DIR, "data")
DB_PATH = os.path.join(DATA_DIR, "repos.db")

os.makedirs(READMES_DIR, exist_ok=True)
os.makedirs(DATA_DIR, exist_ok=True)


def get_gh_token() -> str:
    try:
        res = subprocess.run(
            ["gh", "auth", "token"],
            capture_output=True,
            text=True,
            check=True,
        )
        return res.stdout.strip()
    except Exception:
        return os.environ.get("GITHUB_TOKEN", "")


def load_repos() -> list:
    html_path = os.path.join(SCRIPT_DIR, "index.html")
    if os.path.exists(html_path):
        with open(html_path, "r", encoding="utf-8") as f:
            content = f.read()
        import re

        match = re.search(r"const allRepos = (\[.*?\]);\s+let currentSort", content, re.DOTALL)
        if match:
            return json.loads(match.group(1))
    raise RuntimeError("Could not load allRepos from index.html.")


def init_db() -> sqlite3.Connection:
    conn = sqlite3.connect(DB_PATH)
    conn.execute("PRAGMA journal_mode=WAL;")
    conn.execute("""
        CREATE TABLE IF NOT EXISTS repos (
            id INTEGER PRIMARY KEY,
            db_id INTEGER,
            name TEXT NOT NULL,
            owner TEXT,
            repo_name TEXT,
            url TEXT,
            description TEXT,
            stars INTEGER,
            forks INTEGER,
            language TEXT,
            created_at TEXT,
            updated_at TEXT,
            archived INTEGER,
            readme_filename TEXT,
            readme_size INTEGER,
            readme_text TEXT
        );
    """)
    conn.execute("""
        CREATE VIRTUAL TABLE IF NOT EXISTS repos_fts USING fts5(
            id UNINDEXED,
            name,
            description,
            language,
            readme_text,
            tokenize = 'porter unicode61'
        );
    """)
    return conn


def fetch_readme(repo_name: str, token: str) -> str:
    url = f"https://api.github.com/repos/{repo_name}/readme"
    headers = {
        "User-Agent": "github-best-scraper",
        "Accept": "application/vnd.github.raw+json",
    }
    if token:
        headers["Authorization"] = f"Bearer {token}"

    req = urllib.request.Request(url, headers=headers)
    for attempt in range(1, 4):
        try:
            with urllib.request.urlopen(req, timeout=15) as resp:
                if resp.status == 200:
                    return resp.read().decode("utf-8", errors="replace")
        except urllib.error.HTTPError as e:
            if e.code == 404:
                return ""
            if e.code in (403, 429):
                time.sleep(2)
                continue
            return ""
        except Exception:
            if attempt == 3:
                return ""
            time.sleep(1)
    return ""


def main():
    token = get_gh_token()
    repos = load_repos()
    print(f"Loaded {len(repos)} repositories.")
    print("Scraping READMEs and populating SQLite FTS5 database...")

    conn = init_db()
    conn.execute("DELETE FROM repos;")
    conn.execute("DELETE FROM repos_fts;")
    conn.commit()

    start_time = time.time()
    results = []

    def process_repo(repo):
        name = repo["name"]
        parts = name.split("/")
        owner = parts[0] if len(parts) > 0 else ""
        repo_name = parts[1] if len(parts) > 1 else ""
        filename = f"{str(repo['id']).zfill(4)}_{owner}__{repo_name}.md"
        filepath = os.path.join(READMES_DIR, filename)

        readme_text = ""
        if os.path.exists(filepath) and os.path.getsize(filepath) > 0:
            with open(filepath, "r", encoding="utf-8", errors="replace") as f:
                readme_text = f.read()
        else:
            readme_text = fetch_readme(name, token)
            if readme_text:
                with open(filepath, "w", encoding="utf-8") as f:
                    f.write(readme_text)

        size = len(readme_text.encode("utf-8"))
        return (repo, owner, repo_name, filename, size, readme_text)

    with concurrent.futures.ThreadPoolExecutor(max_workers=20) as executor:
        futures = [executor.submit(process_repo, r) for r in repos]
        completed = 0
        for fut in concurrent.futures.as_completed(futures):
            repo, owner, repo_name, filename, size, readme_text = fut.result()
            results.append((repo, owner, repo_name, filename, size, readme_text))
            completed += 1
            if completed % 100 == 0 or completed == len(repos):
                elapsed = time.time() - start_time
                print(f"Progress: {completed}/{len(repos)} READMEs processed ({elapsed:.1f}s)")

    # Sort by original ID order
    results.sort(key=lambda x: x[0]["id"])

    # Batch insert into DB
    repo_rows = []
    fts_rows = []
    for repo, owner, repo_name, filename, size, readme_text in results:
        repo_rows.append(
            (
                repo["id"],
                repo.get("dbId", repo["id"]),
                repo["name"],
                owner,
                repo_name,
                repo["url"],
                repo.get("desc", ""),
                repo.get("stars", 0),
                repo.get("forks", 0),
                repo.get("lang", "Unknown"),
                repo.get("created", ""),
                repo.get("updated", ""),
                1 if repo.get("archived") else 0,
                filename,
                size,
                readme_text,
            )
        )
        fts_rows.append(
            (
                repo["id"],
                repo["name"],
                repo.get("desc", ""),
                repo.get("lang", "Unknown"),
                readme_text,
            )
        )

    conn.executemany(
        """
        INSERT INTO repos (id, db_id, name, owner, repo_name, url, description, stars, forks, language, created_at, updated_at, archived, readme_filename, readme_size, readme_text)
        VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
    """,
        repo_rows,
    )

    conn.executemany(
        """
        INSERT INTO repos_fts (id, name, description, language, readme_text)
        VALUES (?, ?, ?, ?, ?)
    """,
        fts_rows,
    )
    conn.commit()
    conn.close()

    db_size = os.path.getsize(DB_PATH) / (1024 * 1024)
    print(f"\nDone! SQLite FTS5 database: {DB_PATH} ({db_size:.2f} MB)")
    print(f"Markdown files stored in: {READMES_DIR}")


if __name__ == "__main__":
    main()

File: scrape_readmes.js
const fs = require("node:fs");
const path = require("node:path");
const { execSync } = require("node:child_process");
const { DatabaseSync } = require("node:sqlite");

const SCRIPT_DIR = __dirname || process.cwd();
const READMES_DIR = path.join(SCRIPT_DIR, "readmes");
const DATA_DIR = path.join(SCRIPT_DIR, "data");
const DB_PATH = path.join(DATA_DIR, "repos.db");

if (!fs.existsSync(READMES_DIR)) fs.mkdirSync(READMES_DIR, { recursive: true });
if (!fs.existsSync(DATA_DIR)) fs.mkdirSync(DATA_DIR, { recursive: true });

function getGhToken() {
  try {
    return execSync("gh auth token", { encoding: "utf8", stdio: ["pipe", "pipe", "ignore"] }).trim();
  } catch {
    return process.env.GITHUB_TOKEN || "";
  }
}

const token = getGhToken();
const headers = {
  "User-Agent": "github-best-scraper",
  "Accept": "application/vnd.github.raw+json",
};
if (token) {
  headers.Authorization = `Bearer ${token}`;
}

function loadRepos() {
  const htmlPath = path.join(SCRIPT_DIR, "index.html");
  if (fs.existsSync(htmlPath)) {
    const content = fs.readFileSync(htmlPath, "utf8");
    const match = content.match(/const allRepos = (\[.*?\]);\s+let currentSort/s);
    if (match) {
      return JSON.parse(match[1]);
    }
  }
  throw new Error("Could not find allRepos in index.html. Run generate-and-push first.");
}

function initDb() {
  const db = new DatabaseSync(DB_PATH);
  db.exec(`
    CREATE TABLE IF NOT EXISTS repos (
      id INTEGER PRIMARY KEY,
      db_id INTEGER,
      name TEXT NOT NULL,
      owner TEXT,
      repo_name TEXT,
      url TEXT,
      description TEXT,
      stars INTEGER,
      forks INTEGER,
      language TEXT,
      created_at TEXT,
      updated_at TEXT,
      pushed_at TEXT,
      archived INTEGER,
      readme_filename TEXT,
      readme_size INTEGER,
      readme_text TEXT
    );

    CREATE VIRTUAL TABLE IF NOT EXISTS repos_fts USING fts5(
      id UNINDEXED,
      name,
      description,
      language,
      readme_text,
      tokenize = 'porter unicode61'
    );
  `);

  // Migration check for pushed_at column if older table
  try {
    db.exec("ALTER TABLE repos ADD COLUMN pushed_at TEXT;");
  } catch {
    // Column already exists
  }

  return db;
}

function sanitizeReadme(content) {
  if (!content) return "";
  return content
    .replace(/https:\/\/hooks\.slack\.com\/services\/[A-Za-z0-9_\/]+/g, "https://hooks.slack.com/services_example/T00/B00/XXXX")
    .replace(/https:\/\/discord\.com\/api\/webhooks\/[0-9]+\/[A-Za-z0-9_-]+/g, "https://discord.com/api/webhooks/example/XXXX")
    .replace(/ghp_[A-Za-z0-9]{20,}/g, "ghp_EXAMPLE_TOKEN")
    .replace(/gho_[A-Za-z0-9]{20,}/g, "gho_EXAMPLE_TOKEN")
    .replace(/ghs_[A-Za-z0-9]{20,}/g, "ghs_EXAMPLE_TOKEN")
    .replace(/sk-[A-Za-z0-9_-]{20,}/g, "sk-EXAMPLE_API_KEY")
    .replace(/AKIA[0-9A-Z]{16}/g, "AKIA_EXAMPLE_KEY");
}

async function fetchReadme(repoName) {
  const url = `https://api.github.com/repos/${repoName}/readme`;
  for (let attempt = 1; attempt <= 3; attempt++) {
    try {
      const res = await fetch(url, { headers });
      if (res.status === 200) {
        const text = await res.text();
        return sanitizeReadme(text);
      }
      if (res.status === 404) {
        return "";
      }
      if (res.status === 403 || res.status === 429) {
        console.warn(`Rate limit hit on attempt ${attempt}, waiting 3s...`);
        await new Promise((r) => setTimeout(r, 3000));
        continue;
      }
      return "";
    } catch {
      if (attempt === 3) return "";
      await new Promise((r) => setTimeout(r, 1000));
    }
  }
  return "";
}

async function asyncPool(limit, items, iteratorFn) {
  const ret = [];
  const executing = new Set();
  for (const item of items) {
    const p = Promise.resolve().then(() => iteratorFn(item));
    ret.push(p);
    executing.add(p);
    const clean = () => executing.delete(p);
    p.then(clean, clean);
    if (executing.size >= limit) {
      await Promise.race(executing);
    }
  }
  return Promise.all(ret);
}

async function main() {
  const repos = loadRepos();
  console.log(`Loaded ${repos.length} repositories from index.html.`);

  const db = initDb();

  // Load existing cached records for differential comparison
  const existingMap = new Map();
  try {
    const existingRows = db.prepare("SELECT id, name, description, updated_at, pushed_at, readme_filename, readme_size, readme_text FROM repos").all();
    for (const r of existingRows) {
      existingMap.set(r.name.toLowerCase(), r);
    }
    console.log(`Found ${existingMap.size} existing cached repositories in database.`);
  } catch (e) {
    console.log("No previous database records found. Starting initial sync.");
  }

  let unchangedCount = 0;
  let downloadedCount = 0;
  const startTime = Date.now();
  const processedData = [];

  console.log("Checking for updates (comparing pushed_at / updated_at and description)...");

  await asyncPool(20, repos, async (repo) => {
    const parts = repo.name.split("/");
    const owner = parts[0] || "";
    const repoName = parts[1] || "";
    const safeFilename = `${String(repo.id).padStart(4, "0")}_${owner}__${repoName}.md`;
    const readmeFile = path.join(READMES_DIR, safeFilename);

    const cached = existingMap.get(repo.name.toLowerCase());
    const fileExists = fs.existsSync(readmeFile) && fs.statSync(readmeFile).size > 0;

    const currentPushed = repo.pushedIso || repo.pushed || repo.updatedIso || repo.updated || "";
    const cachedPushed = cached?.pushed_at || cached?.updated_at || "";

    const currentDesc = (repo.desc || "").trim();
    const cachedDesc = (cached?.description || "").trim();

    // Check if both dates and description are unchanged AND file exists
    const isPushedUnchanged = currentPushed && cachedPushed && currentPushed === cachedPushed;
    const isDescUnchanged = currentDesc === cachedDesc;

    let readmeContent = "";
    let isDownloaded = false;

    if (isPushedUnchanged && isDescUnchanged && fileExists) {
      // Unchanged -> Use local cache without network call
      readmeContent = cached?.readme_text || fs.readFileSync(readmeFile, "utf8");
      unchangedCount++;
    } else {
      // Changed or missing -> Download fresh README
      readmeContent = await fetchReadme(repo.name);
      if (readmeContent) {
        fs.writeFileSync(readmeFile, readmeContent, "utf8");
      }
      downloadedCount++;
      isDownloaded = true;
    }

    const readmeSize = Buffer.byteLength(readmeContent, "utf8");
    processedData.push({
      repo,
      owner,
      repoName,
      safeFilename,
      readmeSize,
      readmeContent,
      pushedAt: currentPushed,
      isDownloaded,
    });
  });

  // Re-populate SQLite tables in order of ranking
  processedData.sort((a, b) => a.repo.id - b.repo.id);

  db.exec("DELETE FROM repos; DELETE FROM repos_fts;");

  const insertRepoStmt = db.prepare(`
    INSERT INTO repos (id, db_id, name, owner, repo_name, url, description, stars, forks, language, created_at, updated_at, pushed_at, archived, readme_filename, readme_size, readme_text)
    VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
  `);

  const insertFtsStmt = db.prepare(`
    INSERT INTO repos_fts (id, name, description, language, readme_text)
    VALUES (?, ?, ?, ?, ?)
  `);

  for (const item of processedData) {
    insertRepoStmt.run(
      item.repo.id,
      item.repo.dbId || item.repo.id,
      item.repo.name,
      item.owner,
      item.repoName,
      item.repo.url,
      item.repo.desc || "",
      item.repo.stars || 0,
      item.repo.forks || 0,
      item.repo.lang || "Unknown",
      item.repo.created || "",
      item.repo.updated || "",
      item.pushedAt || "",
      item.repo.archived ? 1 : 0,
      item.safeFilename,
      item.readmeSize,
      item.readmeContent
    );

    insertFtsStmt.run(
      item.repo.id,
      item.repo.name,
      item.repo.desc || "",
      item.repo.lang || "Unknown",
      item.readmeContent
    );
  }

  const elapsed = ((Date.now() - startTime) / 1000).toFixed(2);
  const totalBytes = fs.statSync(DB_PATH).size;

  console.log(`\n=== Synchronization Summary (${elapsed}s) ===`);
  console.log(`• Total Repositories: ${repos.length}`);
  console.log(`• Unchanged (Cached, 0 network requests): ${unchangedCount}`);
  console.log(`• Updated / Newly Downloaded: ${downloadedCount}`);
  console.log(`• SQLite FTS5 Database: ${DB_PATH} (${(totalBytes / 1024 / 1024).toFixed(2)} MB)`);
}

main().catch((err) => {
  console.error("Fatal error:", err);
  process.exit(1);
});
